R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,267 @@
+#' environment outside of the function environment. Sometimes there is no way to understand, just based on static code
+#' analysis, which objects are modified or created by a function call that has side effects. To allow to point to
+#' objects that are modified or created by function calls with side effects, a `@linksto` comment tag was introduced.
+#' This tag enables to restore the reproducible code for an object, with all side effects that do not directly change
+#' this object. Read more about side effects and the usage of `@linksto` tag in [`get_code_dependencies()`] function.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` of length equal to the lenght of `calls_pd` of character vectors of names of objects that are
+#' influenced by `@linksto` tag in a corresponding `call` element of `calls_pd`.
+#' @keywords internal
+#' @noRd
+extract_side_effects <- function(calls_pd) {
+  lapply(
+    calls_pd,
+    function(x) {
+      linksto <- grep("@linksto", x[x$token == "COMMENT", "text"], value = TRUE)
+      unlist(strsplit(sub("\\s*#\\s*@linksto\\s+", "", linksto), "\\s+"))
+    }
+  )
+}
+
+#' @keywords internal
+#' @noRd
+remove_graph_duplicates <- function(graph) {
+  # This function is just a nice to have.
+  # This is not crucial for the implementation.
+  # Purpose: remove duplicates before ":" and after, in each element of list.
+
+  lapply(
+    lapply(graph, paste, collapse = "***"),
+    function(x) {
+      split <- strsplit(x, split = ":", fixed = TRUE)[[1]]
+      deps <- strsplit(split[1], split = "***", fixed = TRUE)[[1]]
+
+      if (length(split) == 1) {
+        unique(deps)
+      } else if (length(split) == 2) {
+        infl <- strsplit(split[2], split = "***", fixed = TRUE)[[1]][-1]
+        c(unique(deps), ":", unique(infl))
+      }
+    }
+  )
+}
+
+# graph_parser ----
+
+#' Return the indexes of calls of code needed to reproduce the object
+#'
+#' @param x The name of the object to return code for.
+#' @param graph A result of `code_graph()`.
+#' @param skip `NULL` or `character` vector. In a recursive call, it is needed to drop parent object to omit dependency
+#' cycles.
+#'
+#' @return `numeric` vector indicating which calls of `graph` are required to build the object passed by name in `x`.
+#'
+#' @keywords internal
+#' @noRd
+graph_parser <- function(x, graph, skip = NULL) {
+  skip <- c(x, skip)
+
+  occurence <-
+    vapply(
+      graph, function(call) {
+        if (":" %in% call) {
+          call <- call[1:(which(":" == call) - 1)]
+        }
+        x %in% call
+      },
+      logical(1)
+    )
+
+  influencers <-
+    unlist(
+      lapply(graph[occurence], function(call) {
+        if (":" %in% call) {
+          call[(which(":" == call) + 1):length(call)]
+        }
```

Comment:
```suggestion
        ind <- match(":", call, nomatch = 0L)
        x %in% call[seq_len(ind + 1L)]
```

Reply:
I actually think this needs to be 

```r
ind <- match("<-", call, nomatch = 0L)
call[(ind + 1L):length(call)]
```

Reply:
I will leave it as 

```r
       if ("<-" %in% call) {
          ind <- match("<-", call, nomatch = 0L)
          call[(ind + 1L):length(call)]
        }
```

Reply:
That's what the `nomatch` is for.

I don't know what names you mean.

Reply:
`call` is a vector of names. if there is no `:` nor `<-` and call looks like this `c("a")` then there are no influencers and I rather return `character(0)` and not TRUE/FALSE

Reply:
the current statement that I think works is here 91327a9bc1592a4d58b104d5f71d4d9b15909121

Reply:
Ooooh, I see the problem.
```
        ind <- match(":", call, nomatch = 0L)
        x %in% call[seq_len(ind + 1L)]
```

I thought you were talking about the "<-" and not the `seq_len`.

You're right, it should be `call[(ind + 1L):length(call)]`.